File: ai/src/game/topology.cpp
#include "game/topology.h"
#include <algorithm>
#include <numeric>
#include <functional>

std::vector<std::vector<int>> zero_adj(int N) {
    return std::vector<std::vector<int>>(N, std::vector<int>(N, 0));
}

// Mirrors shared/topology.ts's own module-private unionFindClasses(): given N nodes and a list of
// pairs to merge, returns each node's equivalence-class index, compressed to a dense 0..M-1 range
// (M = number of distinct classes) in ascending order of each class's lowest original member.
// Internal to merge_boards() (below) - resolves every merge instruction at once, so a chain like
// (0,3)~(1,5) and (1,5)~(2,7) correctly collapses (0,3) and (2,7) into the same node too, even
// though no single instruction names both directly.
static std::vector<int> union_find_classes(int n, const std::vector<std::pair<int,int>>& pairs) {
    std::vector<int> parent(n);
    std::iota(parent.begin(), parent.end(), 0);
    std::function<int(int)> find = [&](int x) -> int {
        while (parent[x] != x) { parent[x] = parent[parent[x]]; x = parent[x]; }
        return x;
    };
    for (auto [a, b] : pairs) {
        int pa = find(a), pb = find(b);
        if (pa != pb) parent[pa] = pb;
    }
    std::vector<int> roots(n);
    for (int i = 0; i < n; i++) roots[i] = find(i);
    std::vector<int> unique_roots = roots;
    std::sort(unique_roots.begin(), unique_roots.end());
    unique_roots.erase(std::unique(unique_roots.begin(), unique_roots.end()), unique_roots.end());
    std::vector<int> root_to_new(n, -1);
    for (size_t i = 0; i < unique_roots.size(); i++) root_to_new[unique_roots[i]] = static_cast<int>(i);
    std::vector<int> node_to_new(n);
    for (int i = 0; i < n; i++) node_to_new[i] = root_to_new[roots[i]];
    return node_to_new;
}

std::pair<RawBoard, std::vector<std::vector<int>>> merge_boards(
    const std::vector<RawBoard>& boards,
    const std::vector<std::pair<std::pair<int,int>, std::pair<int,int>>>& merges) {
    std::vector<int> offset(boards.size(), 0);
    for (size_t i = 1; i < boards.size(); i++)
        offset[i] = offset[i - 1] + static_cast<int>(boards[i - 1].adj.size());
    int total = 0;
    for (const auto& b : boards) total += static_cast<int>(b.adj.size());
    auto g = [&](int b, int local) { return offset[b] + local; };

    std::vector<std::pair<int,int>> pairs;
    for (const auto& m : merges)
        pairs.push_back({ g(m.first.first, m.first.second), g(m.second.first, m.second.second) });
    std::vector<int> node_to_new = union_find_classes(total, pairs);
    int new_n = total == 0 ? 0 : *std::max_element(node_to_new.begin(), node_to_new.end()) + 1;

    std::vector<std::vector<unsigned>> pos(new_n);
    for (size_t b = 0; b < boards.size(); b++)
        for (size_t local = 0; local < boards[b].pos.size(); local++)
            pos[node_to_new[g(static_cast<int>(b), static_cast<int>(local))]] = boards[b].pos[local];

    auto adj = zero_adj(new_n);
    for (size_t b = 0; b < boards.size(); b++) {
        const auto& board_adj = boards[b].adj;
        for (size_t i = 0; i < board_adj.size(); i++)
            for (size_t j = 0; j < board_adj.size(); j++)
                if (board_adj[i][j])
                    adj[node_to_new[g(static_cast<int>(b), static_cast<int>(i))]]
                       [node_to_new[g(static_cast<int>(b), static_cast<int>(j))]] = 1;
    }

    std::vector<std::vector<int>> maps(boards.size());
    for (size_t b = 0; b < boards.size(); b++) {
        maps[b].resize(boards[b].adj.size());
        for (size_t local = 0; local < boards[b].adj.size(); local++)
            maps[b][local] = node_to_new[g(static_cast<int>(b), static_cast<int>(local))];
    }

    std::map<std::string, int> labels;
    for (size_t b = 0; b < boards.size(); b++)
        for (const auto& [addr, local] : boards[b].labels) labels[addr] = maps[b][local];

    return { RawBoard{ std::move(pos), std::move(adj), std::move(labels) }, std::move(maps) };
}

AdjacencyList to_adjacency_list(const std::vector<std::vector<int>>& adj) {
    int N = (int)adj.size();
    AdjacencyList list(N);
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            if (adj[i][j]) list[i].insert(j);
    return list;
}

std::vector<BoardTriangle> find_triangles(const std::vector<std::vector<int>>& adj) {
    int N = (int)adj.size();
    auto adj_list = to_adjacency_list(adj);
    std::vector<BoardTriangle> triangles;
    for (int u = 0; u < N; u++)
        for (int v : adj_list[u]) {
            if (v <= u) continue;
            for (int w : adj_list[v]) {
                if (w <= v) continue;
                if (adj_list[u].count(w)) triangles.push_back(make_board_triangle(u, v, w));
            }
        }
    return triangles;
}

std::vector<BoardSquare> find_squares(const std::vector<std::vector<int>>& adj) {
    int N = (int)adj.size();
    auto adj_list = to_adjacency_list(adj);
    std::vector<BoardSquare> squares;
    for (int p = 0; p < N; p++)
        for (int q = p + 1; q < N; q++) {
            if (adj_list[p].count(q)) continue; // p-q would be an edge, not a diagonal
            std::vector<int> common;
            for (int x : adj_list[p]) if (adj_list[q].count(x)) common.push_back(x);
            for (size_t i = 0; i < common.size(); i++)
                for (size_t j = i + 1; j < common.size(); j++) {
                    int r = std::min(common[i], common[j]);
                    int s = std::max(common[i], common[j]);
                    if (adj_list[r].count(s)) continue; // r-s would be an edge, not a diagonal
                    if (p < r) squares.push_back(make_board_square(p, r, q, s));
                }
        }
    return squares;
}

File: shared/topology.ts
/**
 * Graph-topology utilities operating on plain N×N adjacency matrices (the same representation as
 * `BoardConfig.adj`), independent of any board-specific geometry.
 */

/** An all-zero N×N adjacency matrix - the usual starting point before filling in edges. */
export function zeroAdj(N: number): number[][] {
    return Array.from({ length: N }, () => new Array<number>(N).fill(0));
}

/** An adjacency-list view of a graph: `list[i]` is the set of `i`'s neighbors. */
export type AdjacencyList = Set<number>[];

/** Converts an N×N adjacency matrix into an adjacency list, each node's neighbors stored as a
 * `Set` (not an array) so membership checks - the hot path for both findTriangles/findSquares
 * below - are O(1) instead of O(degree). */
export function toAdjacencyList(adj: number[][]): AdjacencyList {
    const N = adj.length;
    const list: AdjacencyList = Array.from({ length: N }, () => new Set<number>());
    for (let i = 0; i < N; i++)
        for (let j = 0; j < N; j++)
            if (adj[i][j]) list[i].add(j);
    return list;
}

/**
 * Finds every triangle (3 distinct, pairwise-adjacent vertices) in `adj`, each reported exactly
 * once as `[u, v, w]` with `u < v < w`. Converts to an adjacency list first (see toAdjacencyList),
 * then for each vertex `u` only looks at neighbors `v > u`, and for each such `v` only looks at
 * neighbors `w > v` of `v`, checking whether `w` is also a neighbor of `u` via an O(1) set lookup -
 * fixing this increasing order is what guarantees each triangle is found exactly once (via its
 * unique u < v < w labeling), with no separate deduplication pass needed.
 */
export function findTriangles(adj: number[][]): [number, number, number][] {
    const N = adj.length;
    const adjList = toAdjacencyList(adj);
    const triangles: [number, number, number][] = [];
    for (let u = 0; u < N; u++)
        for (const v of adjList[u]) {
            if (v <= u) continue;
            for (const w of adjList[v]) {
                if (w <= v) continue;
                if (adjList[u].has(w)) triangles.push([u, v, w]);
            }
        }
    return triangles;
}

/**
 * Finds every "square" - 4 distinct vertices `a, b, c, d` forming a cycle `a-b-c-d-a` (all 4 cycle
 * edges present) whose two diagonals `a-c` and `b-d` are BOTH absent (a proper induced 4-cycle, not
 * merely 4 vertices of a denser subgraph that happens to contain one) - each reported exactly once
 * as `[a, b, c, d]` in that cycle order.
 *
 * Converts to an adjacency list first (see toAdjacencyList), then for every non-adjacent pair
 * `(p, q)` with `p < q` (a candidate diagonal), finds their common neighbors and, for every pair of
 * common neighbors `(r, s)` that are themselves non-adjacent (the other candidate diagonal), reports
 * the square `p-r-q-s-p`. A square has exactly two diagonals, so this raw scan finds each one twice
 * - once starting from each diagonal - which is resolved by only emitting when `(p, q)` is the
 * lexicographically smaller of the two (`p < min(r, s)`; the two diagonals can never share a vertex,
 * since all 4 square vertices are distinct, so this comparison is never ambiguous).
 */
export function findSquares(adj: number[][]): [number, number, number, number][] {
    const N = adj.length;
    const adjList = toAdjacencyList(adj);
    const squares: [number, number, number, number][] = [];
    for (let p = 0; p < N; p++)
        for (let q = p + 1; q < N; q++) {
            if (adjList[p].has(q)) continue; // p-q would be an edge, not a diagonal
            const common: number[] = [];
            for (const x of adjList[p]) if (adjList[q].has(x)) common.push(x);
            for (let i = 0; i < common.length; i++)
                for (let j = i + 1; j < common.length; j++) {
                    const r = Math.min(common[i], common[j]);
                    const s = Math.max(common[i], common[j]);
                    if (adjList[r].has(s)) continue; // r-s would be an edge, not a diagonal
                    if (p < r) squares.push([p, r, q, s]);
                }
        }
    return squares;
}

/**
 * Union-Find: given `N` nodes and a list of pairs to merge, returns each node's equivalence-class
 * index, compressed to a dense `0..M-1` range (`M` = number of distinct classes) in ascending order
 * of each class's lowest original member. Internal to mergeBoards() (below) - resolves every merge
 * instruction at once, so a chain like `(0,3)~(1,5)` and `(1,5)~(2,7)` correctly collapses `(0,3)`
 * and `(2,7)` into the same node too, even though no single instruction names both directly.
 */
function unionFindClasses(N: number, pairs: [number, number][]): number[] {
    const parent = Array.from({ length: N }, (_, i) => i);
    function find(x: number): number {
        while (parent[x] !== x) { parent[x] = parent[parent[x]]; x = parent[x]; }
        return x;
    }
    for (const [a, b] of pairs) {
        const pa = find(a), pb = find(b);
        if (pa !== pb) parent[pa] = pb;
    }
    const roots = Array.from({ length: N }, (_, i) => find(i));
    const uniqueRoots = [...new Set(roots)].sort((a, b) => a - b);
    const rootToNew = new Map(uniqueRoots.map((r, i) => [r, i]));
    return roots.map(r => rootToNew.get(r)!);
}

/**
 * Combines a list of boards into one, additionally identifying every `([b1, i1], [b2, i2])` pair in
 * `merges` (board index, that board's own local node index) as the same node - every merge is
 * resolved in one batch via unionFindClasses(), not board-by-board, so callers never need to fold
 * boards in one at a time just to keep every merge target "already placed": a merge between two
 * boards that haven't been introduced to each other by any other merge is handled exactly like any
 * other. The merged node keeps whichever input position is encountered first (callers are expected
 * to only merge pairs whose positions already coincide, e.g. two recursive `sierpinskiSimplex`
 * sub-copies sharing a corner). Returns the combined board plus, for each input board (in the same
 * order as `boards`), a map from that board's own local indices to its final index in the combined
 * board - needed by callers that must keep tracking specific nodes (like sierpinskiSimplex's own
 * outer corners) across further merges. `pos` here is opaque per-node data (typically a real
 * position, but never inspected as one) simply carried along through the same merge as `adj`.
 */
export function mergeBoards(
    boards: { pos: number[][]; adj: number[][] }[], merges: [[number, number], [number, number]][],
): { pos: number[][]; adj: number[][]; maps: number[][] } {
    const offset: number[] = new Array(boards.length).fill(0);
    for (let i = 1; i < boards.length; i++) offset[i] = offset[i - 1] + boards[i - 1].pos.length;
    const total = boards.reduce((s, b) => s + b.pos.length, 0);
    const g = (b: number, local: number) => offset[b] + local;

    const nodeToNew = unionFindClasses(total, merges.map(([[b1, i1], [b2, i2]]) => [g(b1, i1), g(b2, i2)]));
    const newN = total === 0 ? 0 : Math.max(...nodeToNew) + 1;

    const pos: number[][] = new Array(newN);
    for (let b = 0; b < boards.length; b++)
        for (let local = 0; local < boards[b].pos.length; local++)
            pos[nodeToNew[g(b, local)]] = boards[b].pos[local];

    const adj = zeroAdj(newN);
    for (let b = 0; b < boards.length; b++) {
        const board = boards[b];
        for (let i = 0; i < board.pos.length; i++)
            for (let j = 0; j < board.pos.length; j++)
                if (board.adj[i][j]) adj[nodeToNew[g(b, i)]][nodeToNew[g(b, j)]] = 1;
    }

    const maps = boards.map((board, b) => board.pos.map((_, local) => nodeToNew[g(b, local)]));
    return { pos, adj, maps };
}
